# CLEANING TOP SONGS

Source: https://www.kaggle.com/edumucelli/spotifys-worldwide-daily-song-ranking/

## Import things

In [2]:
import pandas as pd
import numpy as np

## Load the data

In [3]:
top_songs = pd.read_csv('./data.csv')

## Get to know the data & some initial fix ##

In [4]:
top_songs.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL        Date Region  
0  https://open.spotify.com/track/3AEZUABDXNtecAO...  2017-01-01     ec  
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6...  2017-01-01     ec  
2  https://open.spotify.com/track/3QwBODjSEzelZyV...  2017-01-01     ec  
3  https://open.spotify.com/track/7DM4BPaS7uofFul...  2017-01-01     ec  
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc...  2017-01-01     ec

In [5]:
top_songs.shape

(3441197, 7)

In [6]:
top_songs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3441197 entries, 0 to 3441196
Data columns (total 7 columns):
Position      int64
Track Name    object
Artist        object
Streams       int64
URL           object
Date          object
Region        object
dtypes: int64(2), object(5)
memory usage: 183.8+ MB


COMMENT: I think that the data over all looks good. I need to do some adjustments, based on what I'm goign to use the data for. I might also get rid of some data that I am not going to use.

#### Look at the column Position

In [7]:
top_songs['Position'].isnull().sum()

0

In [8]:
top_songs['Position'].unique()

array([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,  11,  12,  13,
        14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  24,  25,  26,
        27,  28,  29,  30,  31,  32,  33,  34,  35,  36,  37,  38,  39,
        40,  41,  42,  43,  44,  45,  46,  47,  48,  49,  50,  51,  52,
        53,  54,  55,  56,  57,  58,  59,  60,  61,  62,  63,  64,  65,
        66,  67,  68,  69,  70,  71,  72,  73,  74,  75,  76,  77,  78,
        79,  80,  81,  82,  83,  84,  85,  86,  87,  88,  89,  90,  91,
        92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 102, 103, 104,
       105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117,
       118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130,
       131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143,
       144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156,
       157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169,
       170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 18

In [9]:
top_songs['Position'].dtype

dtype('int64')

COMMENT: Nothing really surprising here

#### Look at the column Track Name

In [10]:
top_songs['Track Name'].isnull().sum()

657

In [11]:
top_songs['Track Name'].unique()

array(['Reggaetón Lento (Bailemos)', 'Chantaje',
       'Otra Vez (feat. J Balvin)', ..., '有火', '繼續追', '無期'], dtype=object)

#### Show rows that has NULL values

In [12]:
top_songs[top_songs['Track Name'].isnull()]

Position Track Name Artist  Streams  \
39428          29        NaN    NaN     7362   
39456          57        NaN    NaN     4426   
39463          64        NaN    NaN     4069   
39490          91        NaN    NaN     3238   
39631          32        NaN    NaN     7409   
39666          67        NaN    NaN     4076   
39667          68        NaN    NaN     4059   
39700         101        NaN    NaN     3227   
39962         163        NaN    NaN     2112   
40157         158        NaN    NaN     1905   
61804           5        NaN    NaN    16767   
62004           5        NaN    NaN    19114   
62205           6        NaN    NaN    21390   
62405           6        NaN    NaN    19644   
62606           7        NaN    NaN    16155   
62805           6        NaN    NaN    18439   
113610         11        NaN    NaN    89528   
113633         34        NaN    NaN    56812   
113638         39        NaN    NaN    55701   
113721        122        NaN    NaN    23290   
113814         15        NaN    NaN    87212   
113827         28        NaN    NaN    68460   
113836         37        NaN    NaN    57977   
113970        171        NaN    NaN    23825   
114198        199        NaN    NaN    18868   
114391        192        NaN    NaN    16888   
187828         29        NaN    NaN    71631   
187867         68        NaN    NaN    30024   
187882         83        NaN    NaN    24808   
187913        114        NaN    NaN    19789   
...           ...        ...    ...      ...   
3285870        31        NaN    NaN    82477   
3285885        46        NaN    NaN    65288   
3286126        87        NaN    NaN    29983   
3286321        82        NaN    NaN    30171   
3330726         3        NaN    NaN     2879   
3330731         8        NaN    NaN     2449   
3330749        26        NaN    NaN     1803   
3330787         7        NaN    NaN     2776   
3330797        17        NaN    NaN     2074   
3330814        34        NaN    NaN     1681   
3353257         4        NaN    NaN     5137   
3353259         6        NaN    NaN     4640   
3353279        26        NaN    NaN     2637   
3353296        43        NaN    NaN     1856   
3353332         5        NaN    NaN     4805   
3353333         6        NaN    NaN     4138   
3353359        32        NaN    NaN     2439   
3353374        47        NaN    NaN     1938   
3406398         2        NaN    NaN    13297   
3406402         6        NaN    NaN    11287   
3406420        24        NaN    NaN     7202   
3406435        39        NaN    NaN     5753   
3406601         5        NaN    NaN    12432   
3406612        16        NaN    NaN     9322   
3406633        37        NaN    NaN     6452   
3406651        55        NaN    NaN     5413   
3406883        87        NaN    NaN     3426   
3407072        76        NaN    NaN     3082   
3432077        81        NaN    NaN     3397   
3432383       187        NaN    NaN     1997   

                                                       URL        Date Region  
39428    https://open.spotify.com/track/3RXkboS74UYzN14...  2017-07-20     ec  
39456    https://open.spotify.com/track/4JAyIDXOqNM6qHu...  2017-07-20     ec  
39463    https://open.spotify.com/track/3bVbQvGVIe4n24A...  2017-07-20     ec  
39490    https://open.spotify.com/track/3eFJqPe8VUYrABb...  2017-07-20     ec  
39631    https://open.spotify.com/track/3RXkboS74UYzN14...  2017-07-21     ec  
39666    https://open.spotify.com/track/4JAyIDXOqNM6qHu...  2017-07-21     ec  
39667    https://open.spotify.com/track/3bVbQvGVIe4n24A...  2017-07-21     ec  
39700    https://open.spotify.com/track/3eFJqPe8VUYrABb...  2017-07-21     ec  
39962    https://open.spotify.com/track/6Br5mChPdgQNmLF...  2017-07-22     ec  
40157    https://open.spotify.com/track/6Br5mChPdgQNmLF...  2017-07-23     ec  
61804    https://open.spotify.com/track/1YqcGlCHNquxBhl...  2017-11-09     ec  
62004    https://open.spotify.com/track/1YqcGlCHNquxBhl...  2017-11-10  

In [13]:
# look at unique regions that are missing data
top_songs[top_songs['Track Name'].isnull()]['Region'].unique()
# this code might be wrong...

array(['ec', 'fr', 'ar', 'fi', 'no', 'it', 'lt', 'ph', 'tw', 'nz', 'ee',
       'tr', 'us', 'sv', 'cr', 'de', 'cl', 'jp', 'br', 'hn', 'gt', 'ch',
       'hu', 'ca', 'pe', 'be', 'my', 'dk', 'bo', 'pl', 'at', 'pt', 'se',
       'mx', 'pa', 'uy', 'is', 'es', 'cz', 'ie', 'nl', 'sk', 'co', 'sg',
       'id', 'do', 'lu', 'gb', 'global', 'py', 'au', 'lv', 'gr', 'hk'],
      dtype=object)

I think I have to drop the rows with NULL values for Trackname and Artist.
There is no way I can in any way fill this in. And it is going to be really hard to put together with the analysis from the lyrics, since I do not know what song it is. 
I is 657 songs out of 3,441,197. That is 0.019%...so I still have way more than 95% of my data.
The missing data seem to be distributed over different regions and not only one or a few.

DROP THE ROWS WITH NULL VALUES IN THE END...some rows might be excluded by other reasons.

In [14]:
top_songs['Track Name'].dtype

dtype('O')

COMMENT: I think I might have a need to exclude asia as a region du to different language signs. Will be really hard for the lyrics part. Might have to exclude other regions as well, based on that.

#### Look at the column Artist

In [15]:
top_songs['Artist'].isnull().sum()

657

In [16]:
top_songs['Artist'].unique()

array(['CNCO', 'Shakira', 'Zion & Lennox', ..., 'E-Kids', 'Jane Zhang',
       '帶菌者'], dtype=object)

In [17]:
top_songs['Artist'].dtype

dtype('O')

COMMENT: I think I might have a need to exclude asia as a region du to different language signs. Will be really hard for the lyrics part. Might have to exclude other regions as well, based on that.

#### Look at the column Streams

In [18]:
top_songs['Streams'].isnull().sum()

0

In [19]:
top_songs['Streams'].dtype

dtype('int64')

COMMENT: Looks good. No NULL-values and they are all integer. Hurray!

#### Look at the column Date

In [20]:
top_songs['Date'].isnull().sum()

0

In [21]:
top_songs['Date'].dtype

dtype('O')

In [22]:
# converting into datetime-format
top_songs['Date'] = pd.to_datetime(top_songs['Date'])

In [23]:
top_songs['Date'].dtype

dtype('<M8[ns]')

In [24]:
top_songs['Date'].min()

Timestamp('2017-01-01 00:00:00')

In [25]:
top_songs['Date'].max()

Timestamp('2018-01-09 00:00:00')

In [26]:
# I do want to remove the 9 days in 2018. I only want data for 2017. That is one of my limitations. 

In [27]:
# make new column for year. Then drop all rows from 2018
top_songs['Year'] = top_songs['Date'].dt.year

In [28]:
top_songs.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL       Date Region  Year  
0  https://open.spotify.com/track/3AEZUABDXNtecAO... 2017-01-01     ec  2017  
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6... 2017-01-01     ec  2017  
2  https://open.spotify.com/track/3QwBODjSEzelZyV... 2017-01-01     ec  2017  
3  https://open.spotify.com/track/7DM4BPaS7uofFul... 2017-01-01     ec  2017  
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc... 2017-01-01     ec  2017

In [29]:
#  make new data frame with only data for year 2017
top_songs = top_songs[top_songs['Year'] != 2018]

In [30]:
top_songs['Date'].max()

Timestamp('2017-12-31 00:00:00')

In [31]:
# make new column for month. 
top_songs['Month'] = top_songs['Date'].dt.month

In [32]:
# make new column for day. 
top_songs['Day'] = top_songs['Date'].dt.day

In [33]:
top_songs.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL       Date Region  Year  \
0  https://open.spotify.com/track/3AEZUABDXNtecAO... 2017-01-01     ec  2017   
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6... 2017-01-01     ec  2017   
2  https://open.spotify.com/track/3QwBODjSEzelZyV... 2017-01-01     ec  2017   
3  https://open.spotify.com/track/7DM4BPaS7uofFul... 2017-01-01     ec  2017   
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc... 2017-01-01     ec  2017   

   Month  Day  
0      1    1  
1      1    1  
2      1    1  
3      1    1  
4      1    1

COMMENT: When all the data put together is in one data frame the Date column has to be transformed into datetime-format again (that format is not saved I think i the csv file). I might also like to make separat columns for year, month and day separatly. 

#### Look at the column Region (will be renamed to Country)

In [34]:
# Rename the column Region to Country
top_songs.rename(columns={'Region': 'Country'}, inplace=True)

In [35]:
top_songs.columns

Index(['Position', 'Track Name', 'Artist', 'Streams', 'URL', 'Date', 'Country',
       'Year', 'Month', 'Day'],
      dtype='object')

In [36]:
top_songs['Country'].isnull().sum()

0

In [37]:
top_songs['Country'].unique()

array(['ec', 'fr', 'ar', 'fi', 'no', 'it', 'lt', 'ph', 'tw', 'nz', 'ee',
       'tr', 'us', 'sv', 'cr', 'de', 'cl', 'jp', 'br', 'hn', 'gt', 'ch',
       'hu', 'ca', 'pe', 'be', 'my', 'dk', 'bo', 'pl', 'at', 'pt', 'se',
       'mx', 'pa', 'uy', 'is', 'es', 'cz', 'ie', 'nl', 'sk', 'co', 'sg',
       'id', 'do', 'lu', 'gb', 'global', 'py', 'au', 'lv', 'gr', 'hk'],
      dtype=object)

In [38]:
np.sort(top_songs['Country'].unique())

array(['ar', 'at', 'au', 'be', 'bo', 'br', 'ca', 'ch', 'cl', 'co', 'cr',
       'cz', 'de', 'dk', 'do', 'ec', 'ee', 'es', 'fi', 'fr', 'gb',
       'global', 'gr', 'gt', 'hk', 'hn', 'hu', 'id', 'ie', 'is', 'it',
       'jp', 'lt', 'lu', 'lv', 'mx', 'my', 'nl', 'no', 'nz', 'pa', 'pe',
       'ph', 'pl', 'pt', 'py', 'se', 'sg', 'sk', 'sv', 'tr', 'tw', 'us',
       'uy'], dtype=object)

In [39]:
top_songs['Country'].dtype

dtype('O')

On Kaggle someone already created a dictonary with the real names of the regions (countries). Very convenient!<BR />
I will put it in the read me file under info about data or something similar<BR />
https://www.kaggle.com/mrncst/simple-exploratory-analysis-of-spotify-songs<BR />
I do not want to change my names however (a short version is esier to work with).<BR />
Why do I need to know this. Bc I do want to group countries into bigger "regions" based on continents.

'ar':'Argentina'<BR />
'at':'Austria'<BR />
'au':'Australia'<BR />
'be':'Belgium'<BR />
'bo':'Bolivia'<BR />
'br':'Brazil'<BR />
'ca':'Canada'<BR />
'ch':'Switzerland'<BR />
'cl':'Chile'<BR />
'co':'Columbia'<BR />
'cr':'CostaRica'<BR />
'cz':'CzechRepublic'<BR />
'de':'Germany'<BR />
'dk':'Denmark'<BR />
'do':'DominicanRepublic'<BR />
'ec':'Ecuador'<BR />
'ee':'Estonia'<BR />
'es':'Spain'<BR />
'fi':'Finland'<BR />
'fr':'France'<BR />
'gb':'UnitedKingdom'<BR />
'global':'World'<BR />
'gr':'Greece'<BR />
'gt':'Guatemala'<BR />
'hk':'HongKong'<BR />
'hn':'Honduras'<BR />
'hu':'Hungary'<BR />
'id':'Indonesia'<BR />
'ie':'Ireland'<BR />
'is':'Iceland'<BR />
'it':'Italy'<BR />
'jp':'Japan'<BR />
'lt':'Lithuania'<BR />
'lu':'Luxemborg'<BR />
'lv':'Latvia'<BR />
'mx':'Mexico',<BR /> 
'my':'Malaysia'<BR />
'nl':'Netherlands'<BR />
'no':'Norway'<BR />
'nz':'NewZealand'<BR />
'pa':'Panama'<BR />
'pe':'Peru',<BR />
'ph':'Philippines'<BR />
'pl':'Poland'<BR />
'pt':'Portugal'<BR />
'py':'Paraguay'<BR />
'se':'Sweden'<BR />
'sg':'Singapore'<BR />
'sk':'Slovakia'<BR />
'sv':'ElSalvador'<BR />
'tr':'Turkey'<BR />
'tw':'Taiwan'<BR />
'us':'USA'<BR />
'uy':'Uruguay'<BR />

In [40]:
# GROUP THE CONTRIES INTO REGIONS. Make a new column that indicated what bigger region (based on continent) 
# a country in the column Country belongs to. 
# Name the "new column" Region

**South America (8)**<BR />
'ar':'Argentina'<BR />
'bo':'Bolivia'<BR />
'br':'Brazil'<BR />
'co':'Columbia'<BR />
'ec':'Ecuador'<BR />
'pe':'Peru',<BR />
'py':'Paraguay'<BR />
'uy':'Uruguay'<BR />

**Europe (26)**<BR />
'at':'Austria'<BR />
'be':'Belgium'<BR />
'ch':'Switzerland'<BR />
'cz':'CzechRepublic'<BR />
'de':'Germany'<BR />
'dk':'Denmark'<BR />
'ee':'Estonia'<BR />
'es':'Spain'<BR />
'fi':'Finland'<BR />
'fr':'France'<BR />
'gb':'UnitedKingdom'<BR />
'gr':'Greece'<BR />
'hu':'Hungary'<BR />
'ie':'Ireland'<BR />
'is':'Iceland'<BR />
'it':'Italy'<BR />
'lt':'Lithuania'<BR />
'lu':'Luxemborg'<BR />
'lv':'Latvia'<BR />
'nl':'Netherlands'<BR />
'no':'Norway'<BR />
'pl':'Poland'<BR />
'pt':'Portugal'<BR />
'se':'Sweden'<BR />
'sk':'Slovakia'<BR />
'tr':'Turkey'<BR />

**Oceania (2)**<BR />
'au':'Australia'<BR />
'nz':'NewZealand'<BR />


**Nothe America & Caribean (9)**<BR />
'ca':'Canada'<BR />
'cr':'CostaRica'<BR />
'do':'DominicanRepublic'<BR />
'gt':'Guatemala'<BR />
'hn':'Honduras'<BR />
'mx':'Mexico',<BR /> 
'pa':'Panama'<BR />
'sv':'ElSalvador'<BR />
'us':'USA'<BR />

**Asia (7)**<BR />
'hk':'HongKong'<BR />
'id':'Indonesia'<BR />
'jp':'Japan'<BR />
'my':'Malaysia'<BR />
'ph':'Philippines'<BR />
'sg':'Singapore'<BR />
'tw':'Taiwan'<BR />

**Other (1)**<BR />
'global':'World'<BR />

In [41]:
# Create new column for Region (based on continent)
def country_to_region(country):
    sa = ['ar', 'bo', 'br', 'co', 'ec', 'pe', 'py', 'uy']
    eu = ['at', 'be', 'ch', 'cz', 'de', 'dk', 'ee', 'es', 'fi', 'fr', 
          'gb', 'gr', 'hu', 'ie', 'is', 'it', 'lt', 'lu', 'lv', 'nl', 
          'no', 'pl', 'pt', 'se', 'sk', 'tr']
    oc = ['au', 'nz' ]
    na = ['ca', 'cr', 'do', 'gt', 'hn', 'mx', 'pa', 'sv', 'us']
    asi = ['hk', 'id', 'jp', 'my', 'ph', 'sg', 'tw']
    
    if country in sa:
        return 'sa'
    elif country in eu:
        return 'eu'
    elif country in oc:
        return 'oc'
    elif country in na:
        return 'na'
    elif country in asi:
        return 'as'
    else:
        return 'world'

top_songs['Region'] = top_songs['Country'].apply(country_to_region)

In [42]:
top_songs['Region'].unique()

array(['sa', 'eu', 'as', 'oc', 'na', 'world'], dtype=object)

COMMENT: I might want to focus more on one of the regions (based on continent). The column Global is what I think it it, it can be used for feature engineering. You might want to see if a country is deviating a lot from global/average for example.

**Look at missing value again for the columns Track Name and Artist**

In [43]:
top_songs['Track Name'].isnull().sum()

657

In [44]:
top_songs['Artist'].isnull().sum()

657

In [45]:
top_songs.shape

(3356486, 11)

Drop rows with missing values in any column (we know that there are only missing values in the columns Track Name and Artist)

In [46]:
top_songs = top_songs.dropna(axis=0, how='any')

In [47]:
top_songs.shape

(3355829, 11)

In [48]:
# Look that we have successfully dropped 657 rows. (It also tells us that all the missing values that existed in the 
# column Track Name corresponded ti missing values in the column Artist.)
3356486-3355829

657

In [ ]:
# Keep in mind that some positions will be missing after dropping rows

Change the order of the columns

In [49]:
top_songs.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL       Date Country  Year  \
0  https://open.spotify.com/track/3AEZUABDXNtecAO... 2017-01-01      ec  2017   
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6... 2017-01-01      ec  2017   
2  https://open.spotify.com/track/3QwBODjSEzelZyV... 2017-01-01      ec  2017   
3  https://open.spotify.com/track/7DM4BPaS7uofFul... 2017-01-01      ec  2017   
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc... 2017-01-01      ec  2017   

   Month  Day Region  
0      1    1     sa  
1      1    1     sa  
2      1    1     sa  
3      1    1     sa  
4      1    1     sa

In [50]:
top_songs = top_songs[['Position', 'Track Name', 'Artist', 'Streams', 'URL', 'Date', 'Year', 'Month', 'Day', 'Country', 'Region']]

In [51]:
top_songs.head()

Position                  Track Name         Artist  Streams  \
0         1  Reggaetón Lento (Bailemos)           CNCO    19272   
1         2                    Chantaje        Shakira    19270   
2         3   Otra Vez (feat. J Balvin)  Zion & Lennox    15761   
3         4                Vente Pa' Ca   Ricky Martin    14954   
4         5                      Safari       J Balvin    14269   

                                                 URL       Date  Year  Month  \
0  https://open.spotify.com/track/3AEZUABDXNtecAO... 2017-01-01  2017      1   
1  https://open.spotify.com/track/6mICuAdrwEjh6Y6... 2017-01-01  2017      1   
2  https://open.spotify.com/track/3QwBODjSEzelZyV... 2017-01-01  2017      1   
3  https://open.spotify.com/track/7DM4BPaS7uofFul... 2017-01-01  2017      1   
4  https://open.spotify.com/track/6rQSrBHf7HlZjtc... 2017-01-01  2017      1   

   Day Country Region  
0    1      ec     sa  
1    1      ec     sa  
2    1      ec     sa  
3    1      ec     sa  
4    1      ec     sa

## Save to a csv

In [52]:
# load the data to a .csv
top_songs.to_csv('top_songs_clean.csv')